### PR TITLE
Bring back error attachments

### PR DESCRIPTION
**Pull request by clpolet**
Please don't merge until backend is ready :)

    * Error attachment is a log on its own
    * Nil or empty filename is replaced by auto generated ones
    * Bring back the crash callback to get attachments for an error report
    * Send error attachments right after the related crash log
    * Don't buffer error attachments
    * Make MSAbstractLog & MSLogWithProperties public since MSErrorAttachmentLog is also public
    * Move public models to a dedicated "model" folder
    * Fix incoherencies on the type properties of models
    * Fix Unit tests
    * Add Unit tests for error attachments
    * Updated puppet & SasquatchObJC with error attachment delegate

### CODE REVIEW

File: MobileCenterCrashes/MobileCenterCrashes/Internals/Model/MSAbstractErrorLog.h [hunk 1]
```diff
@@ -1,14 +1,17 @@
 #import <Foundation/Foundation.h>
 #import "MobileCenter+Internal.h"
 
+#import "MSLogWithPropertiesInternal.h"
+#import "MobileCenter+Internal.h"
```
Comment (jaelim-ms) [inline MobileCenterCrashes/MobileCenterCrashes/Internals/Model/MSAbstractErrorLog.h]: Duplicate import

File: MobileCenterCrashes/MobileCenterCrashes/Internals/Model/MSErrorAttachmentLogInternal.h [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ * Error log identifier to attach this log to.
+ */
+@property(nonatomic, copy) NSString *errorId;
+
+/**
+ * Checks if the object's values are valid.
+ *
+ * return YES, if the object is valid
```
Comment (jaelim-ms) [inline MobileCenterCrashes/MobileCenterCrashes/Internals/Model/MSErrorAttachmentLogInternal.h]: `@return` and period at the end of the sentence. Only for this file. I already addressed this in a separate PR.

File: MobileCenterCrashes/MobileCenterCrashes/MSCrashes.mm [hunk 3]
```diff
@@ -175,17 +181,23 @@ + (void)notifyWithUserConfirmation:(MSUserConfirmation)userConfirmation {
...
-
-      // TODO (attachmentWithCrashes): Bring this back when the backend supports attachment for Crashes.
-      //      [log setErrorAttachment:[crashes.delegate attachmentWithCrashes:crashes forErrorReport:report]];
+      attachments = [crashes.delegate attachmentWithCrashes:crashes forErrorReport:report];
     } else {
       MSLogDebug([MSCrashes logTag], @"attachmentWithCrashes is not implemented");
     }
 
-    // Send log to log manager.
+    // First send crash to log manager.
```
Comment (jaelim-ms) [inline MobileCenterCrashes/MobileCenterCrashes/MSCrashes.mm]: `crash` -> `crash log`?

File: MobileCenterCrashes/MobileCenterCrashes/MSCrashes.mm [hunk 4]
```diff
@@ -175,17 +181,23 @@ + (void)notifyWithUserConfirmation:(MSUserConfirmation)userConfirmation {
...
     [crashes.logManager processLog:log withPriority:crashes.priority andGroupID:crashes.groupID];
+
+    // Then send attachements to log manager.
```
Comment (jaelim-ms) [inline MobileCenterCrashes/MobileCenterCrashes/MSCrashes.mm]: `attachments` -> `error attachment logs`?

File: MobileCenterCrashes/MobileCenterCrashes/Model/MSErrorAttachmentLog.h [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+/**
+ * Create an attachment with a given filename and text.
+ *
+ * @param filename The filename the attachment should get. If nil will get an automatically generated filename.
+ * @param text The attachment text.
+ *
+ * @return An instance of `MSErrorAttachmentLog`.
+ */
++ (MSErrorAttachmentLog *)attachmentWithText:(NSString *)text filename:(NSString *)filename;
+/**
```
Comment (jaelim-ms) [inline MobileCenterCrashes/MobileCenterCrashes/Model/MSErrorAttachmentLog.h]: New line.

File: MobileCenterCrashes/MobileCenterCrashes/Internals/MSErrorAttachmentLog.m [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+- (void)encodeWithCoder:(NSCoder *)coder {
+  [super encodeWithCoder:coder];
+  [coder encodeObject:self.attachmentId forKey:kMSId];
+  [coder encodeObject:self.errorId forKey:kMSErrorId];
+  [coder encodeObject:self.contentType forKey:kMSContentType];
+  [coder encodeObject:self.filename forKey:kMSFileName];
+  [coder encodeObject:self.data forKey:kMSData];
+}
+
+#pragma mark - Public Interface
```
Comment (jaelim-ms) [inline MobileCenterCrashes/MobileCenterCrashes/Internals/MSErrorAttachmentLog.m]: These should be in a separate class. These are utility methods, shouldn't be in *Log classes, we don't expose those internal classes. Android is using `ErrorAttachments` class, iOS can use `MSErrorAttachments` with static methods.

Comment (clpolet) [inline MobileCenterCrashes/MobileCenterCrashes/Internals/MSErrorAttachmentLog.m]: Why are those not exposed publicly?

**Comment by clpolet** [comment]
@jaelim-ms MSDevice is in `MobileCenter/Model` moved all public model there to match crash module. I think it makes sens but let me know otherwise.

**Comment by clpolet** [comment]
Closing this one for now, will reopen when backend will be ready.

**Pull request closed without merging**